File: Algorithms/031.NextPermutation.java
public class Solution {
    public void nextPermutation(int[] num) {
        for (int i=num.length-1; i > 0; i--) {
            if (num[i] <= num[i-1]) {
                continue;
            }
            for (int j=num.length-1; j >= i; j--) {
                if (num[j] > num[i-1]) {
                    int temp = num[i-1];
                    num[i-1] = num[j];
                    num[j] = temp;
                    for(int p=i, q=num.length-1; p < q; p++, q--) {
                        temp = num[p];
                        num[p] = num[q];
                        num[q] = temp;
                    }
                    return;
                }
            }
        }
        for(int p=0, q=num.length-1; p < q; p++, q--) {
            int temp = num[p];
            num[p] = num[q];
            num[q] = temp;
        }
    }
}

File: Algorithms/cpp/031.NextPermutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        for (int i=nums.size()-1; i > 0; i--) {
            if (nums[i] <= nums[i-1]) {
                continue;
            }
            for (int j=nums.size()-1; j >= i; j--) {
                if (nums[j] > nums[i-1]) {
                    int temp = nums[i-1];
                    nums[i-1] = nums[j];
                    nums[j] = temp;
                    for(int p=i, q=nums.size()-1; p < q; p++, q--) {
                        int temp = nums[p];
                        nums[p] = nums[q];
                        nums[q] = temp;
                    }
                    return;
                }
            }
        }
        for(int p=0, q=nums.size()-1; p < q; p++, q--) {
            int temp = nums[p];
            nums[p] = nums[q];
            nums[q] = temp;
        }
    }
};
